### PR TITLE
Develop

### LINKED ISSUE #39: gitignore
GitHub brepo should contain only C++ files.
No .sln, .pdb, .exe, etc

### LINKED ISSUE #43: Try/catch logic
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L11-L22
Throwing exceptions for immediately catching is pointless.
```cpp
if (size < 0) size = 1;
```
implies the same logic.

### LINKED ISSUE #44: new(0)
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L13
What if `size == 0`?

### LINKED ISSUE #46: C++-style
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L33
NULL is just a C macro for zero. Use nullptr instead

### LINKED ISSUE #50: Excess nesting
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L67
```cpp
if (this == &value) return;
```

### LINKED ISSUE #52: Duplicate code
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L81
This is almost the same as find function. It could
```cpp
return find(value) != nullptr;
```

### LINKED ISSUE #53: C++-style
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L99
`NULL` is a C keyword/macro (usually `#define NULL 0`). C++ analogue is `nullptr`

### LINKED ISSUE #54: Zero capacity allocation
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L107-L113
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L30-L40

- `1 != 0` so there will be no realloction.
- even if it would, two by zero is still zero.

### LINKED ISSUE #57: Destructor
It would be better if you assign the pointer to `nullptr`
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L25-L28

### LINKED ISSUE #58: Unsafe function
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L90
I can use this function to get an access to your set.
Thus I am able e.g. to make your set unsorted w/o triggering any other interface.

### LINKED ISSUE #59: Bad asymptotics
https://github.com/CherryCobbler/lab17/blob/379196047a093c72a3de8a8422289de5f53b1ec7/Project5/Source.cpp#L90
O(size), should be O(log(size))

### PR DESCRIPTION
**Pull request by CherryCobbler**
close #59 
close #58 
close #53
close #54 
close #52 
close #50 
close #46 
close #43 
close #57 
close #39 
close #44

### CODE REVIEW

File: Project5/Header.h [hunk 1]
```diff
@@ -6,22 +6,23 @@
...
 	template <typename T>
 	friend std::ostream& operator<< (std::ostream& out, const Set<T>& value);
 	bool find(const T& value) const;//needed for a task
-	T* find_it(const T& value) const;//needed for the insert function
+	const T* find_it(const T& value) const;//needed for the insert function
 	size_t set_size() const;
 	void clear();
-	int erase(const T& value);//If returned -2 - set empty, if -1 - element not found
+	const T* erase(const T* it);
+	const T* set_begin() const;
```
Comment (ghost) [inline Project5/Header.h:26]: What is that?

File: Project5/Source.cpp [hunk 2]
```diff
@@ -8,34 +8,29 @@ Set<T>::Set()
...
-		std::cout << "Error: " << thr <<"! You can't create a Set with "<<nsize<< " elements! Created a Set for one element.";
-		size = 1;
-		array = new T[1];
 	}
 }
 template<typename T>
 Set<T>::~Set()
 {
 	delete[] array;
+	array = NULL;
```
Comment (ghost) [inline Project5/Source.cpp:22]:
```cpp
array = nullptr;
```

File: Project5/Source.cpp [hunk 3]
```diff
@@ -56,47 +51,40 @@ Set<T>::Set(const Set& value)
...
-		for (int i = 0; i < busy; i++)
-		{
-			array[i] = value.array[i];
-		}
-	}
+	if (this == &value) return *this;
+	size = value.size;
+	busy = value.busy;
+	delete[] array;
+	array = new T[size];
```
Comment (Just-Nik) [inline Project5/Source.cpp:63]:
```cpp
try{
...
array = new T[size];
...
}
catch(std::bad_alloc){throw;}

File: Project5/Source.cpp [hunk 4]
```diff
@@ -109,44 +97,17 @@ void Set<T>::clear()
...
-		return thr;
-	}
-	busy--;
-	return rez;
+const T* Set<T>::erase(const T* it)
+{
+	//Will be soon
+}
+template<typename T>
+const T* Set<T>::set_begin() const
```
Comment (ghost) [inline Project5/Source.cpp:108]: pls rename to begin()

File: Project5/Source.cpp [hunk 5]
```diff
@@ -8,34 +8,29 @@ Set<T>::Set()
 template<typename T>
 Set<T>::Set(const size_t nsize)
 {
-	try
+	if (nsize == 0) array = nullptr;
+	else
 	{
-		if (nsize < 0) throw -3;
+		array = new T[nsize];
```
Comment (Just-Nik) [inline Project5/Source.cpp:14]:
```cpp
try{
...
array = new T[size];
...
}
catch(std::bad_alloc){throw;}

**Review by ghost — COMMENTED** [review]
#53  not fixed. All other remarks in the comments.

**Comment by marcfedorow** [comment]
#39 is not closed actually

**Pull request merged**